docs/javascript/render-functions.md

```diff
@@ -0,0 +1,292 @@
+
+## Ruby Helper Functions with Render Functions
+
+### 1. react_component
+
+The `react_component` helper is used for rendering a single React component. It accepts various return types from render functions:
+
+```ruby
+# Basic usage with a component
+<%= react_component("MyComponent", props: { name: "John" }) %>
+
+# With a render function that returns JSX
+<%= react_component("MyRenderFunction", props: { name: "John" }) %>
+
+# With a render function that returns an object with renderedHtml
+<%= react_component("MyRenderFunction", props: { name: "John" }) %>
+
+# With a render function that returns a Promise
+<%= react_component("MyAsyncRenderFunction", props: { name: "John" }) %>
+```
+
+#### When to use:
+
+- When you need to render a single component
+- When you're rendering client-side only
+- When your render function returns a single HTML string
+
+#### Not suitable for:
+
+- When your render function returns an object with multiple HTML strings
+- When you need to insert content in different parts of the page
+
+### 2. react_component_hash
+
+The `react_component_hash` helper is used when your render function returns an object with multiple HTML strings. It allows you to place different parts of the rendered output in different parts of your layout.
+
+```ruby
+# With a render function that returns an object with multiple HTML properties
+<% helmet_data = react_component_hash("HelmetComponent", props: {
+  title: "My Page",
+  description: "Page description"
+}) %>
+
+<% content_for :head do %>
+  <%= helmet_data["title"] %>
+  <%= helmet_data["metaTags"] %>
+<% end %>
+
+<div class="main-content">
+  <%= helmet_data["componentHtml"] %>
+</div>
+```
+
+#### When to use:
+
+- When your render function returns multiple HTML strings in an object
+- When you need to insert rendered content in different parts of your page
+- For SEO-related rendering like meta tags and title tags
+- When working with libraries like React Helmet
+
+#### Not suitable for:
+
+- Simple component rendering
+- Client-side only rendering (always uses server rendering)
+
+#### Requirements:
+
+- The render function MUST return an object
+- The object MUST include a `componentHtml` key
+- All other keys are optional and can be accessed in your Rails view
+
+## Best Practices Based on the Tests
+
+1. **Always Use Objects for Multiple HTML Parts**: If you need multiple HTML strings, return an object with named properties, and use `react_component_hash` to access them.
+
+2. **Don't Return Raw HTML Strings**: Tests show that returning a raw HTML string directly causes errors - either use JSX or wrap in a `{ renderedHtml: '...' }` object.
+
+3. **Async Functions Work with Objects**: For async render functions, you can return both strings and objects (with or without a `renderedHtml` property).
```

Comment:
The three sections above seem redundant as they state the same thing as lines 73-79.